test/core/tsi/s2a/s2a_record_protocol_test.cc

```diff
@@ -245,12 +244,259 @@ static void test_create_crypter_success(TLSCiphersuite ciphersuite) {
+  if (ciphersuite == TLS_CHACHA20_POLY1305_SHA256_ciphersuite) {
+    GPR_ASSERT(status == GRPC_STATUS_UNIMPLEMENTED);
+    int correct_error_message =
+        strcmp(error_details, S2A_CHACHA_POLY_UNIMPLEMENTED);
+    GPR_ASSERT(correct_error_message == 0);
+
+    // Cleanup.
+    s2a_crypter_destroy(crypter);
+    gpr_free(error_details);
+    grpc_core::Delete<grpc_channel>(channel);
+    return;
+  }
+  GPR_ASSERT(status == GRPC_STATUS_OK);
+  GPR_ASSERT(error_details == nullptr);
+
+  const size_t test_plaintext_size = 7;
+  uint8_t test_plaintext[test_plaintext_size] = "123456";
+  size_t record_allocated_size =
+      test_plaintext_size + s2a_max_record_overhead(crypter);
+  uint8_t* record =
+      (uint8_t*)gpr_malloc(record_allocated_size * sizeof(uint8_t));
+  size_t record_size;
+  grpc_status_code encrypt_status =
+      s2a_encrypt(crypter, test_plaintext, test_plaintext_size - 1, record,
+                  record_allocated_size, &record_size, &error_details);
+  GPR_ASSERT(encrypt_status == GRPC_STATUS_OK);
+  GPR_ASSERT(record_size == expected_message_size(test_plaintext_size - 1));
+
+  bool correct_encrypted_record = check_encrypt_record(
+      ciphersuite, record, record_size, nullptr, 0, nullptr, 0, &error_details);
+  GPR_ASSERT(correct_encrypted_record);
+  GPR_ASSERT(error_details == nullptr);
+
+  // Cleanup.
+  s2a_crypter_destroy(crypter);
+  grpc_core::Delete<grpc_channel>(channel);
+  gpr_free(record);
+}
+
+static void test_encrypt_three_records(TLSCiphersuite ciphersuite) {
+  s2a_crypter* crypter = nullptr;
+  grpc_channel* channel = grpc_core::New<grpc_channel>();
+  char* error_details = nullptr;
+  grpc_status_code status =
+      setup_crypter(ciphersuite, channel, &crypter, &error_details);
+  if (ciphersuite == TLS_CHACHA20_POLY1305_SHA256_ciphersuite) {
+    GPR_ASSERT(status == GRPC_STATUS_UNIMPLEMENTED);
+    int correct_error_message =
+        strcmp(error_details, S2A_CHACHA_POLY_UNIMPLEMENTED);
+    GPR_ASSERT(correct_error_message == 0);
+
+    // Cleanup.
+    s2a_crypter_destroy(crypter);
+    gpr_free(error_details);
+    grpc_core::Delete<grpc_channel>(channel);
+    return;
+  }
+  GPR_ASSERT(status == GRPC_STATUS_OK);
+  GPR_ASSERT(error_details == nullptr);
+
+  const size_t test_plaintext_one_size = 7;
+  uint8_t test_plaintext_one[test_plaintext_one_size] = "123456";
+  size_t record_one_allocated_size =
+      test_plaintext_one_size + s2a_max_record_overhead(crypter);
+  uint8_t* record_one =
+      (uint8_t*)gpr_malloc(record_one_allocated_size * sizeof(uint8_t));
+  size_t record_one_size;
+  grpc_status_code encrypt_one_status = s2a_encrypt(
+      crypter, test_plaintext_one, test_plaintext_one_size - 1, record_one,
+      record_one_allocated_size, &record_one_size, &error_details);
+  GPR_ASSERT(encrypt_one_status == GRPC_STATUS_OK);
+  GPR_ASSERT(record_one_size ==
+             expected_message_size(test_plaintext_one_size - 1));
+
+  const size_t test_plaintext_two_size = 10;
```

Comment:
I've added a utility function that captures the encryption and the subsequent `GPR_ASSERT`s. I didn't wrap the initialization of the plaintext and record into the utility function, however, because all the records are needed for verification after all encryption is done. 

If you had something else in mind, let me know!

Reply:
SG, thanks!